sql/src/main/java/io/crate/execution/IncrementalPageBucketReceiver.java

```diff
@@ -45,43 +46,66 @@
     private final T state;
     private final AtomicInteger remainingUpstreams;
     private final CompletableFuture<Iterable<Row>> processingFuture = new CompletableFuture<>();
+    private final Executor executor;
     private final Streamer<?>[] streamers;
 
     private volatile Throwable lastThrowable = null;
     private final BatchIterator<Row> lazyBatchIterator;
+    private CompletableFuture<?> currentlyAccumulating;
+    private final Function<Bucket, Void> accumulateRowsFunction;
 
     public IncrementalPageBucketReceiver(Collector<Row, T, Iterable<Row>> collector,
                                          RowConsumer rowConsumer,
+                                         Executor executor,
                                          Streamer<?>[] streamers,
                                          int upstreamsCount) {
         this.state = collector.supplier().get();
         this.accumulator = collector.accumulator();
         this.finisher = collector.finisher();
+        this.executor = executor;
         this.streamers = streamers;
         this.remainingUpstreams = new AtomicInteger(upstreamsCount);
         lazyBatchIterator = CollectingBatchIterator.newInstance(
             () -> {},
             t -> {},
             () -> processingFuture);
         rowConsumer.accept(lazyBatchIterator, null);
-    }
-
-    @Override
-    public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageResultListener pageResultListener) {
-        pageResultListener.needMore(!isLast);
-        synchronized (state) {
+        accumulateRowsFunction = (Bucket rows) -> {
             try {
                 for (Row row : rows) {
                     accumulator.accept(state, row);
                 }
             } catch (Throwable e) {
                 lastThrowable = e;
+                throw e;
+            }
+            return null;
+        };
+    }
+
+    @Override
+    public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageResultListener pageResultListener) {
+        pageResultListener.needMore(!isLast);
+
+        synchronized (state) {
```

Comment:
Maybe use something other than `state` to synchronize? It gives the impression that concurrent access to `state` is being prevented, but this actually seems to happen by using the future mechanism to chain multiple operations.

Reply:
yeah, I actually thought that's a feature as it points out that `state` shouldn't be assumed as thread-safe

Reply:
pushed a fixup

Reply:
Fair point. But it confused me initially and I wanted to point out that the state synchronization doesn't protect the state within the function because the synchronization block is exited early. Then I figured how it actually works.

Maybe a comment along the lines of `We make sure only one accumulation operation runs at a time because the state is not thread-safe` would help to clear things up

Reply:
Yeah, fair enough. I can see how that can be confusing. Pushed another fixup to add a comment and signal that the completable future (which is initially null) is guarded by a different field synchronisation wise.